### DISCUSSION TITLE
A small bug about fuel inflow velocity and question about How do I get the desired run data and output it to a plt file

### DISCUSSION THREAD

**Discussion by grabtheNS** [AMReX-Combustion/PeleLMeX#440] [General]
Hello,in Exec/Production/DiffBunsen2D/pelelmex_prob.H ，line222
```
    case 1: // Fuel inflow
    {
      // The speed in the fuel inflow has a parabolic profile - the fuel tube is
      // supposed ideal and the speed is set at 0 at the limits of the tube.
      a = (-1.5) * prob_parm.V_fu / (prob_parm.Xf * prob_parm.Xf);
      c = 1.5 * prob_parm.V_fu;
      s_ext[VELY] = a * x[0] * x[0] + c;
```

This parabolic velocity curve form, when integrated for a given inlet velocity prob_parm.V_fu, does not directly yield the volume flow rate in the form of the average velocity multiplied by the exit area. In actual simulations, the calculation of the mass flow rate of carbon exiting the furnace is very important. I would recommend using this fully developed velocity curve form [laminarsmoke](https://github.com/acuoci/laminarSMOKE/blob/master/run/tutorials/ToroFlames/F3/laminarBuoyantPimpleSMOKE/0/U)  ,which, after my calculations, yields the volume flow rate equal to the average velocity multiplied by the exit area.
`      s_ext[VELY] = 2 * prob_parm.V_fu * (1 - (x[0] * x[0]) / (prob_parm.Xf * prob_parm.Xf));
`
![output](https://github.com/user-attachments/assets/8eb85928-b29f-40b8-9ad5-991ec564ca92)

And I have a quesiton about how do I get the desired run data and output it to a plt file.For example, in the process of calculating soot, I want to get the rate of PAH consumption per grid, which is calculated in the source code, how do I pack them into an array and then output them in the Plt file

**Comment by baperry2** [comment]
Regarding the velocity profiles, if an alternative specification is more suitable for your case, go for it. We expose these functions to users so you can make adjustments as needed for your cases. In general, specifying the value from the analytical profile at the cell center will lead to some numerical error on the average volumetric inflow rate, but this should converge as resolution is added. The ideal solution consistent with the finite volume numerics is probably to analytically integrate the inflow profile over the inflow face of each cell and apply that value.

To plot variables beyond the standard state and derived variables, you can create a user-defined derived variable that contains the values you need as (briefly described in the [documentation](https://amrex-combustion.github.io/PeleLMeX/manual/html/LMeXControls.html#pelelmex-derived-variables). That may or may not work for your needs depending on whether the necessary data is available in the arguments to that function. Another option would be to write an additional plot file by adding a call to [amrex::WriteMultiLevelPlotfile](https://amrex-codes.github.io/amrex/docs_html/IO.html) in the source code where the quantity gets computed. If you're a bit more specific about what you want to plot and where it is computed, we may be able to provide a more tailored suggestion.

**Reply by grabtheNS** [reply]
I would like to express my sincere gratitude for your assistance. I have recently been getting acquainted with the data structures and usage of Amrex.It is quite simple to store specific values and output the values of a field in OpenFOAM. However, I am not familiar with how to perform similar operations in PeleLMex. I want to store the value of dimerRate for each grid calculation process and output it in a plot file.How should I modify this part of the code?

在[Source/Soot/SootData.H](https://github.com/AMReX-Combustion/PelePhysics/blob/development/Source/Soot/SootData.H) row 120 
```
    // Compute the dimerization rate
    const amrex::Real dimerRate = sr->dimerRate(T, xi_PAH);
```

**Reply by baperry2** [reply]
In this case `dimerRate` is not a field variable, but a local variable that exists only in compute kernels. It looks like the computation is pretty simple, so it may not be too hard to add a user-defined derived variable to recompute this quantity and dump into a plot file using the capability I linked to above.

**Reply by grabtheNS** [reply]
Yes, for dimerRate, as you suggested I managed to get the output of its Plt variable, but now I have a new problem, I want to calculate the average diffusion coefficient D_diff for each mesh, and I checked the source code, and it seems to be very complicated to use this method for this variable, do you have any other suggestions?

**Reply by baperry2** [reply]
Diffusion coefficients are available as default derived variables. You can output diffusion coefficients to a plot file using `amr.derive_plot_vars = diffcoeff`. Note that the variables output are not the raw diffusion coefficient (D with units of m^2/s), but rather are the effective coefficient used in the algorithm, `beta_i = rho*Di *Wbar/W_i`

**Reply by grabtheNS** [reply]
Thanks for your patient and professional answer. In soot, variables such as k_sg is a very important metric to assess the growth rate, it is neither simple can be recalculated by known results like dimerRate nor in preset output variables like D_diff, it relies on the calculation of most local variables in PeleSuit, and after a few days of attempts on my part to do so by recalculating seems to be having difficulty annoying, do you have any relevant suggestions? 

/PelePhysics/Source/Soot/SootData.H  row133
```
    amrex::Real k_sg = 0.;
    amrex::Real k_ox = 0.;
    amrex::Real k_o2 = 0.;
    // Compute the species reaction source terms into omega_src
    sr->chemicalSrc(T, surf, xi_n, moments, k_sg, k_ox, k_o2, omega_src);
```
